Builder pattern, XPath parsing, deprecate insecure, change initialization

remingtonc opened the pull request:
**EDIT**: The scope of this pull request is getting out of hand, but it's all good stuff. Think it's ready to merge. Bumping to `v1.0.0` due to major changes.

Refactors the internal codebase to be more extensible, particularly with respect to potential deviations in how XPaths are handled and with regard to security, and includes XPath parsing which can handle filters.

* Usage is not simply `XRClient(...)`, changed to builder pattern with `ClientBuilder`. I am especially interested in your opinions around how this looks.

Original:
```python
client = XRClient('127.0.0.1:57400', 'demo', 'demo', credentials='ems.pem', tls_server_override='ems.cisco.com', credentials_from_file=True)
capabilities = client.capabilities()
print(capabilities)
```

New:
```python
from cisco_gnmi import ClientBuilder
client = ClientBuilder(
    '127.0.0.1:9339'
).set_os(
    'IOS-XR'
).set_secure_from_file(
    'ems.pem'
).set_ssl_target_override(
    'ems.cisco.com'
).set_call_authentication(
    'admin',
    'its_a_secret'   
).construct()
capabilities = client.capabilities()
print(capabilities)
```

It's very difficult to mask direct interaction with the gRPC libraries without this sort of initialization, or having tons of optional init arguments. It becomes important once security gets involved. The work done here I will likely attempt to percolate into our other libraries too, so it sets a small precedence on how we initialize clients.

* Wrapper classes now override `parse_xpath_to_gnmi_path` in order to appropriately handle origins etc.
* Default gNMI port changed to 9339 per [reserved_port.md](https://github.com/openconfig/gnoi/blob/master/docs/reserved_port.md).

**EDITed past here**

* Almost forgot, also revamps XPath parsing to actually parse filters etc. correctly into gNMI paths. Borrows the xml module's XPath regex and works around its results. This was very critical.
* Deprecates insecure. This is expressly against gNMI spec and is pretty trivial with the ability to get the certificate from the target.
* Uses channel credentials instead of passing metadata with every call.
* `Client` and child classes simply initialize on top of a gRPC channel instead of any connectivity logic in the constructor.

This work was derived from adding an IOS XE wrapper, but having challenges with SSL & friends. This work in the meantime is still worth discussing/merging.

Resolves #5 
Basis for #10 and #9

remingtonc committed:
Refactor to chained initialization

remingtonc committed:
Fix cryptography usage

remingtonc committed:
Default to create a secure service

remingtonc committed:
Specify service in base constructor

remingtonc committed:
Move xpath parsing to impl class

remingtonc committed:
XPath filter -> Path key support

remingtonc committed:
Improve XPath parsing doc

remingtonc committed:
Use new init format in classdoc

remingtonc committed:
Fix incorrect init arg

remingtonc committed:
Fix parse_xpath_to_gnmi_path refs

remingtonc committed:
Client XPath/encoding fixes

remingtonc committed:
Fix certificate loading logic

remingtonc committed:
Include py2-ipaddress for Python 2 support

remingtonc committed:
Remove IOS XE functionality

remingtonc committed:
Set default port to 9339 (IANA reserved by OC)

remingtonc committed:
Scope cert parsing for filesystem as well

remingtonc committed:
Touch up documentation

remingtonc commented:
ping

skkumaravel commented on src/cisco_gnmi/base.py:
```diff
@@ -0,0 +1,201 @@
+        return self
+
+    def as_insecure(self, channel_options=None, compression=None):
+        logging.warning(
```
Needs a function comment

skkumaravel reviewed (changes requested):
Looks good, couple comments:

1. Interesting way to get the CA name, I like it! Is that considered secure, or should that be a flag that is allowed to be passed? The reason I am asking is because by passing the CA name and the pem file, you are ensuring that the pem file is the right pem file and didn't get tampered with.

2. You should also add an example of insecure and secure with file name. Little confused on how to use all the different ways to do an init by just reading the code, especially the secure one.

remingtonc committed:
Document and override as_insecure

remingtonc committed:
Add documentation around initialization

remingtonc committed:
Fix as_insecure override

remingtonc committed:
Clean up dependencies

remingtonc commented:
@skkumaravel Addressed comments, README.md has more detail now. Regarding the CA name - in the case of `ems.pem` we still need to set the override as `ems.cisco.com` regardless of where it is loaded - so yes it is still useful even locally. However I am also not an expert in this situation.

remingtonc committed:
Address non-black pylint

remingtonc committed:
Revert "Address non-black pylint"

This reverts commit c330b2a35ba58d190e36e815400b23762b0e2b41.

remingtonc committed:
Address non-black pylint

remingtonc committed:
Quick formatting fix

ziausyed commented on README.md:
```diff
@@ -19,6 +19,148 @@ It is *highly* recommended that users of the library learn [Google Protocol Buff
+
+
 ### XRClient
 `XRClient` inherets from `Client` and provides several wrapper methods which aid with IOS XR-specific behaviors of the gNMI implementation. These are `delete_xpaths`, `get_xpaths`, `set_json`, and `subscribe_xpaths`. These methods make several assumptions about what kind of information will be supplied to them in order to simplify usage of the gNMI RPCs.
```
Do you want to specify what assumptions are made?

remingtonc replied:
Not necessarily, the method doc and parameters themselves kind of embody what those assumptions are.

remingtonc replied:
Called out the assumptions being in the doc.

ziausyed reviewed (commented):
very cool changes.

skkumaravel commented on README.md:
```diff
@@ -19,6 +19,148 @@ It is *highly* recommended that users of the library learn [Google Protocol Buff
+client = Client(
+    '127.0.0.1:9339'
+).as_secure(
+    'ems.pem',
```
In this example, how is the CA determined?

remingtonc replied:
I quite honestly do not know. This is a good exercise in security.

skkumaravel reviewed (approved):
README helps a lot! Looks really good now, just had the one question..

remingtonc committed:
Realize builder pattern, deprecate insecure, use channel authentication

remingtonc commented:
@brh55, despite approving, brought up that what was presented wasn't exactly a builder pattern. He was correct, and I almost brushed it off, but rewrote and now is drastically more builder...and pushed in some extra changes while rolling...please re-review.

PS I'm sorry about wild PR scope creep.

remingtonc committed:
Run black formatting

remingtonc committed:
Add set_target method

remingtonc committed:
Fix IOS-XR example doc

remingtonc committed:
IOS-XR -> IOS XR

remingtonc committed:
Remove reference to _gen_metadata

remingtonc commented:
Resolves #21
